### PR TITLE
Alamofire 5: Encodable Request Parameters

### PR DESCRIPTION
**Pull request by jshier**
### Issue Link :link:
#2181

### Goals :soccer:
This PR adds the ability to make requests using `Encodable` types as parameters. It includes the `ParameterEncoder` protocol, as well as `JSONParameterEncoder` and `URLEncodedFormParameterEncoder`.

### Implementation Details :construction:
Implementation is relatively straightforward, except for the `URLEncodedFormEncoder` that was added. That type is based on a combination of [Vapor's `URLEncodedFormEncoder`](https://github.com/vapor/url-encoded-form) and [Flight School's DIY Codable Kit](https://github.com/Flight-School/Codable-DIY-Kit). It's a bit nasty, since form encoding has special requirements for nested types.

### Testing Details :mag:
Tests have been added for the new form encoder, but other tests still need to be added.

- [x] Add more tests.
- [x] Add inline docs.

### CODE REVIEW

File: Tests/ParameterEncoderTests.swift [hunk 1]
```diff
@@ -0,0 +1,451 @@
...
+
+        // When
+        let result = Result<String> { try encoder.encode(parameters) }
+
+        // Then
+        let escaped = String(repeating: "%E4%B8%80%E4%BA%8C%E4%B8%89%E5%9B%9B%E4%BA%94%E5%85%AD%E4%B8%83%E5%85%AB%E4%B9%9D%E5%8D%81",
+                             count: repeatedCount)
+        let expected = "chinese=\(escaped)"
+        XCTAssertEqual(result.value, expected)
+    }
```
Comment (cnoon) [inline Tests/ParameterEncoderTests.swift]: Looking good @jshier! With the complexity of all the new logic for url encoding, I wanted to make sure we have all the right tests. I enabled code coverage and found we're only at 43% on the `ParameterEncoder` file. This strikes me as complex enough that we should really shoot for 100% code coverage. Otherwise I'm afraid we'll end up with issues that could be very difficult to track down and will be easy to cause other regressions. Could you take a stab at bumping it up?

File: Source/ParameterEncoder.swift [hunk 2]
```diff
@@ -0,0 +1,863 @@
...
+    /// Encode the provided `Encodable` parameters into `request`.
+    ///
+    /// - Parameters:
+    ///   - parameters: The `Encodable` parameter value.
+    ///   - request:    The `URLRequest` into which to encode the parameters.
+    /// - Returns:      A `URLRequest` with the result of the encoding.
+    /// - Throws:       An `Error` when encoding fails. For Alamofire provided encoders, this will be an instance of
+    ///                 `AFError.parameterEncoderFailed` with an associated `ParameterEncoderFailureReason`.
+    func encode<Parameters: Encodable>(_ parameters: Parameters?, into request: URLRequest) throws -> URLRequest
+}
```
Comment (cnoon) [inline Source/ParameterEncoder.swift]: This is such an awesome design! It should eliminate all the issues we've faced with parameter encoding flexibility in the past. Excellent work!

File: Source/Alamofire.swift [hunk 3]
```diff
@@ -38,34 +38,51 @@ public enum AF {
     ///   - encoding:   The `ParameterEncoding`, `URLEncoding.default` by default.
     ///   - headers:    The `HTTPHeaders`, `nil` by default.
     /// - Returns:      The created `DataRequest`.
-    @discardableResult
```
[the remark this answers is not in the document]
Comment (jshier) [reply Source/Alamofire.swift]: When wouldn't you use the produced `*Request`s? It just seemed odd to me that we were allowing them to be discarded when `Session`'s API doesn't allow that, and discarding them doesn't allow attaching response handlers, which makes the request rather pointless.

File: Source/ParameterEncoding.swift [hunk 4]
```diff
@@ -24,23 +24,6 @@
...
-    case patch   = "PATCH"
-    case delete  = "DELETE"
-    case trace   = "TRACE"
-    case connect = "CONNECT"
-}
-
-// MARK: -
-
 /// A dictionary of parameters to apply to a `URLRequest`.
 public typealias Parameters = [String: Any]
```
[the remark this answers is not in the document]
Comment (jshier) [reply Source/ParameterEncoding.swift]: For existing users. While we'll want to remove it eventually, I didn't want to break every AF4 user's existing codebase. Transitioning to `Encodable` parameters may not be feasible for them.

File: Source/ParameterEncoder.swift [hunk 5]
```diff
@@ -0,0 +1,863 @@
...
+        func encode(_ string: String) -> String {
+            switch self {
+            case .percentEscaped: return string.replacingOccurrences(of: " ", with: "%20")
+            case .plusReplaced: return string.replacingOccurrences(of: " ", with: "+")
+            }
+        }
+    }
+
+    /// Internal `URLEncodedFormEncoder` error.
+    public enum Error: Swift.Error {
```
[the remark this answers is not in the document]
Comment (jshier) [reply Source/ParameterEncoder.swift]: I'll update the docstring, it's the "internal" error type of the encoder, not internal to AF.

File: Source/ParameterEncoder.swift [hunk 6]
```diff
@@ -0,0 +1,863 @@
...
+/// `application/x-www-form-urlencoded; charset=utf-8`.
+///
+/// There is no published specification for how to encode collection types. By default, the convention of appending
+/// `[]` to the key for array values (`foo[]=1&foo[]=2`), and appending the key surrounded by square brackets for
+/// nested dictionary values (`foo[bar]=baz`) is used. Optionally, `ArrayEncoding` can be used to omit the
+/// square brackets appended to array keys.
+///
+/// `BoolEncoding` can be used to configure how boolean values are encoded. The default behavior is to encode
+/// `true` as 1 and `false` as 0.
+open class URLEncodedFormParameterEncoder: ParameterEncoder {
```
[the remark this answers is not in the document]
Comment (jshier) [reply Source/ParameterEncoder.swift]: Mostly, I used this name to match a similar encoder from Vapor and because, technically, that's the proper name for this encoding: URL encoded form. I'll think more about the naming though, as it is verbose.

File: Tests/ParameterEncoderTests.swift (hunk 1 above)
Comment (jshier) [reply Tests/ParameterEncoderTests.swift]: I've been able to bump it up to 70%. Manually testing the various nested containers looks tricky.

**Comment by jshier** [comment]
@cnoon Coverage is over 80% now, which is about as good as we'll get. This is ready for another review.

File: Source/Alamofire.swift (hunk 3 above)
[the remark this answers is not in the document]
Comment (cnoon) [reply Source/Alamofire.swift]: Oh I follow now. You are correct. You're thinking that you always chain or capture the request to then attach validations or response handlers to it, which is absolutely correct. Therefore, it doesn't make sense to use `@discardableResult` on the top-level APIs. As long as we maintain `@discardableResult` on all the chaining APIs, then I think this makes complete sense.

File: Source/ParameterEncoder.swift (hunk 6 above)
[the remark this answers is not in the document]
Comment (cnoon) [reply Source/ParameterEncoder.swift]: Where are you finding that the "proper" name is URL encoded form? Everyone certainly refers to is as URL encoding, and the "form" portion simply lives in the header and seems to be legacy as per the original implementation.

File: Source/ParameterEncoding.swift (hunk 4 above)
[the remark this answers is not in the document]
Comment (cnoon) [reply Source/ParameterEncoding.swift]: Might be worth at least deprecating it in the initial release. Could also be a 5.1 change.

**Comment by cnoon** [comment]
Why can we only hit 80% coverage? Just curious because some of the worst bugs lie in those difficult to test edge cases.

**Comment by jshier** [comment]
You can take a look at the coverage and see if there's anything you find important. I'll take another look tonight. Basically, I don't find it worth the effort to hit 100% for lines that aren't relevant to the feature.

File: Source/ParameterEncoding.swift (hunk 4 above)
[the remark this answers is not in the document]
Comment (jshier) [reply Source/ParameterEncoding.swift]: I'd rather wait so users can move to 5 without immediate deprecation warnings. We can deprecate before we decide to remove.

File: Source/ParameterEncoder.swift (hunk 6 above)
[the remark this answers is not in the document]
Comment (jshier) [reply Source/ParameterEncoder.swift]: To me `URLEncoded` implies more about it being in the URL than `URLEncodedForm`.

**Comment by jshier** [comment]
Coverage is at 83.2% for the `ParameterEncoder.swift` file. The remaining things are stuff like the convenience static properties for encoders with different settings, `localizedDescription`, some unused bits from code I cribbed from elsewhere, and nil encoding scenarios I couldn't trigger (since they actually all get channeled through a single implementation). Let me know if there's anything in particular you'd like to see tested.

**Comment by jshier** [comment]
@cnoon Added a few more tests so coverage is now 91.3%, which is as high as it's going to get.
